### ISSUE TITLE
Adding dataset Tiny-Imagenet

### ISSUE THREAD

**Issue by towzeur** [pytorch/vision#6127]
### 🚀 The feature

Hello, 

I would like to contribute to torchvision by providing a implementation of Tiny-Imagenet dataset.

home : https://www.kaggle.com/c/tiny-imagenet
paper : http://vision.stanford.edu/teaching/cs231n/reports/2015/pdfs/yle_project.pdf
zip : http://cs231n.stanford.edu/tiny-imagenet-200.zip

This challenge is part of Stanford Class CS 231N.
Label Classes and Bounding Boxes are provided

**details:**
classes : 200
image_size : 64x64x3
bbox : x0, y0, x1, y1 for each image
train split : 100 000  (500 per class) 
val split : 10 000 (50 per class)
test split : 10 000 (50 per class)

### Motivation, pitch

Note: the original test split doesn't have targets and bboxes.
Thus, in this implementation, I used the val split when passing `train=True`.

Features:
- fast loading by creating numpy files (npy/*.npy) from the raw folder/image datasets
- can leverage bbox

Structure:
```
root
├───tiny-imagenet-200.zip
├───tiny-imagenet-200
│   ├───npy <-- generated
│   │       ├───test_bboxes.npy
│   │       ├───test_data.npy
│   │       ├───test_targets.npy
│   │       ├───train_bboxes.npy
│   │       ├───train_data.npy
│   │       ├───train_targets.npy
│   ├───test
│   ├───train
│   ├───val
│   ├───words.txt
│   └───wnids.txt
```

Here the implementation: 
https://github.com/towzeur/vision/commit/a67feb569361f440fd48ed492183de8bd8f6b585

**Comment by datumbox** [comment]
@towzeur Thanks for offering to help!

Let's get back to you on this. We are in the process of migrating to a new API, so it's unclear if at this point we will add more datasets on the old one. I've recorded your proposal on the RFC at #3562 so that we won't forget it. BTW note that putting the standard `torchvision.datasets.ImageFolder`on top of a folder that contains the extract data should work for Tiny Imagenet. We've been using it internally to check quickly models, so that's a quick workaround until it's properly added.

cc @NicolasHug

**Comment by towzeur** [comment]
Thanks for your response !

Yes we can use torchvision.datasets.ImageFolder but only for the "train" split 
which have the required structure (same semantic images grouped in the same dir, 
```
1. tiny-imagenet-200\train\n01443537\images
...
200. tiny-imagenet-200\train\n12267677\images
```

Problem:
ImageFolder would provide a unique class for all the images in the val and test set
because images are grouped in a single "image" folder regardless of their semantic label
```
- tiny-imagenet-200\val\images 
- tiny-imagenet-200\test\images
```

So one have to parse the `val_annotations.txt` to retrieve the labels of val images.
Note that their are annotations files as well for each category of the train set
```
1. tiny-imagenet-200\train\n01443537\n01443537_boxes.txt
...
200. tiny-imagenet-200\train\n12267677\n12267677_boxes.txt
```
Note 2. test split doesn't have an annotations file so it's a unsupervised set.

In my implementation, I parse all annotations files (train and val) as it can fix the previous problem
and it further provide bboxes of the object in interest for each image.

I also think that having numpy file leads to faster loading as it bypass the 
file structure parsing and it can fit directly in memory.
Combined, all the numpy files (images, targets, bbox)
of the 'val' and 'train' split weights 1.26 Go.

**Comment by datumbox** [comment]
@towzeur I agree that adding the Tiny ImageNet is worth it. We need to see if this can be done on the current dataset API or on the new one. If it's the latter, it might take a bit more time so I would advice to parse the val annotations as you described to make it usable with `ImageFolder`. Concerning your point on the lack of annotations on test images, I believe that's something that was considered on the new Datasets API which is under development. 

cc @NicolasHug @pmeier
